### PR TITLE
Add com.mojang.Minecraft

### PR DESCRIPTION
**Pull request by Kekun**
This should hopefully work but I didn't manage to test it via installation, etc.

### CODE REVIEW

File: com.mojang.Minecraft.json [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      "--extra-data=Minecraft.jar:e8d90b2637b141b556670d8460d317aadb6e4f2e644e519995c428800b25f212:280212:280212:http://s3.amazonaws.com/Minecraft.Download/launcher/Minecraft.jar"
+  ],
+  "modules": [
+      {
+          /* Needed by Minecraft 1.8.2 and up. */
+          "name": "xrandr",
+          "sources": [
+              {
+                  "type": "git",
+                  "url": "git://anongit.freedesktop.org/xorg/app/xrandr"
```
Comment (TingPing) [inline com.mojang.Minecraft.json]: Don't build from git.

File: com.mojang.Minecraft.json [hunk 2]
```diff
@@ -0,0 +1,65 @@
+{
+  /* GTK+ is needed by JavaFX. */
+  "app-id": "com.mojang.Minecraft",
+  "runtime": "org.freedesktop.Sdk.Extension.openjdk9",
```
Comment (TingPing) [inline com.mojang.Minecraft.json]: This should be in `"sdk-extensions"` not set as the runtime.

File: com.mojang.Minecraft.json [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      "--persist=.minecraft",
+      /* X11 access */
+      "--socket=x11",
+      /* OpenGL access */
+      "--device=dri",
+      /* PulseAudio access */
+      "--socket=pulseaudio",
+      /* Network access */
+      "--share=network",
+      "--extra-data=Minecraft.jar:e8d90b2637b141b556670d8460d317aadb6e4f2e644e519995c428800b25f212:280212:280212:http://s3.amazonaws.com/Minecraft.Download/launcher/Minecraft.jar"
```
Comment (TingPing) [inline com.mojang.Minecraft.json]: If you have a recent `flatpak` there is a new source type that cleans this up. https://github.com/flathub/com.discordapp.Discord/blob/master/com.discordapp.Discord.json#L33-L39

File: com.mojang.Minecraft.json (hunk 2 above)
Comment (Alexander-Wilms) [reply com.mojang.Minecraft.json]: And it needs a module which copies the JDK: https://github.com/flathub/org.octave.Octave/blob/master/org.octave.Octave.json#L320

**Comment by Kekun** [comment]
Thanks! I applied all the fixes you suggested, used the GNOME SDK/Platform and simplified the indentation.

**Comment by Kekun** [comment]
The Minecraft launcher prints thins in both the terminal and its window:
```
Bootstrap (v5)
Current time is Nov 25, 2017, 12:42:12 PM
System.getProperty('os.name') == 'Linux'
System.getProperty('os.version') == '4.13.10-1-default'
System.getProperty('os.arch') == 'amd64'
System.getProperty('java.version') == '9'
System.getProperty('java.vendor') == 'Oracle Corporation'
System.getProperty('sun.arch.data.model') == '64'

Downloading: https://s3.amazonaws.com/Minecraft.Download/launcher/launcher.pack.lzma
Exception: javax.net.ssl.SSLException: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
Downloading: https://s3.amazonaws.com/Minecraft.Download/launcher/launcher.pack.lzma (try 2/10)
Exception: javax.net.ssl.SSLException: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
Downloading: https://s3.amazonaws.com/Minecraft.Download/launcher/launcher.pack.lzma (try 3/10)
Exception: javax.net.ssl.SSLException: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
Downloading: https://s3.amazonaws.com/Minecraft.Download/launcher/launcher.pack.lzma (try 4/10)
Exception: javax.net.ssl.SSLException: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
Downloading: https://s3.amazonaws.com/Minecraft.Download/launcher/launcher.pack.lzma (try 5/10)
Exception: javax.net.ssl.SSLException: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
Downloading: https://s3.amazonaws.com/Minecraft.Download/launcher/launcher.pack.lzma (try 6/10)
Exception: javax.net.ssl.SSLException: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
Downloading: https://s3.amazonaws.com/Minecraft.Download/launcher/launcher.pack.lzma (try 7/10)
Exception: javax.net.ssl.SSLException: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
Downloading: https://s3.amazonaws.com/Minecraft.Download/launcher/launcher.pack.lzma (try 8/10)
Exception: javax.net.ssl.SSLException: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
Downloading: https://s3.amazonaws.com/Minecraft.Download/launcher/launcher.pack.lzma (try 9/10)
Exception: javax.net.ssl.SSLException: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
Downloading: https://s3.amazonaws.com/Minecraft.Download/launcher/launcher.pack.lzma (try 10/10)
Exception: javax.net.ssl.SSLException: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
Unable to download remote file. Check your internet connection/proxy settings.
FATAL ERROR: net.minecraft.bootstrap.FatalBootstrapError: Unable to download while being forced
	at net.minecraft.bootstrap.Bootstrap.execute(Bootstrap.java:95)
	at net.minecraft.bootstrap.Bootstrap.main(Bootstrap.java:381)


Please fix the error and restart.
```

But if I run it out of the sandbox and then in the sanbox it works. I suppose it is because it downloads the required files in `~/.minecraft` and then is fine. Removing `~/.minecraft` and starting in the sandbox fails anew.

**Comment by Kekun** [comment]
Here is the output out of the sandbox, for comparison:
```
Bootstrap (v5)
Current time is Nov 25, 2017 1:51:53 PM
System.getProperty('os.name') == 'Linux'
System.getProperty('os.version') == '4.13.10-1-default'
System.getProperty('os.arch') == 'amd64'
System.getProperty('java.version') == '1.8.0_144'
System.getProperty('java.vendor') == 'Oracle Corporation'
System.getProperty('sun.arch.data.model') == '64'

Downloading: https://s3.amazonaws.com/Minecraft.Download/launcher/launcher.pack.lzma
Got reply in: 411ms
Downloaded 1077,2kb in 0s at 1388,3kb/s
Renaming /home/kekun/.minecraft/launcher.pack.lzma.new to /home/kekun/.minecraft/launcher.pack.lzma
Renamed successfully.
Reversing LZMA on /home/kekun/.minecraft/launcher.pack.lzma to /home/kekun/.minecraft/launcher.pack
Unpacking /home/kekun/.minecraft/launcher.pack to /home/kekun/.minecraft/launcher.jar
Cleaning up /home/kekun/.minecraft/launcher.pack
Starting launcher.
[13:51:57 INFO]: Minecraft Launcher 1.6.84-j (through bootstrap 5) started on linux...
[13:51:57 INFO]: Current time is Nov 25, 2017 1:51:57 PM
[13:51:57 INFO]: System.getProperty('os.name') == 'Linux'
[13:51:57 INFO]: System.getProperty('os.version') == '4.13.10-1-default'
[13:51:57 INFO]: System.getProperty('os.arch') == 'amd64'
[13:51:57 INFO]: System.getProperty('java.version') == '1.8.0_144'
[13:51:57 INFO]: System.getProperty('java.vendor') == 'Oracle Corporation'
[13:51:57 INFO]: System.getProperty('sun.arch.data.model') == '64'
[13:51:57 INFO]: proxy == DIRECT
[13:51:57 INFO]: Refreshing local version list...
[13:51:57 INFO]: Refreshing remote version list...
[13:51:57 INFO]: Refresh complete.
[13:51:57 INFO]: Loaded 0 profile(s); selected '(Default)'
```

**Comment by Kekun** [comment]
Also, font are ugly, any idea how to fix that? I remember that one between the proprietary and the free JREs work well regarding font, but I dont' remember which one.
capture d ecran de 2017-11-25 13-53-53

**Comment by Kekun** [comment]
The network just doesn't seem to work in the sandbox (I pre-downloaded the launcher out of the sanbox, then ran it in the sandbox and got this when trying to log in):
capture d ecran de 2017-11-25 13-56-20

File: com.mojang.Minecraft.appdata.xml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    </p>
+  </description>
+  <url type="homepage">https://minecraft.net/</url>
+  <screenshots>
+    <screenshot type="default">
+      <image>https://lh3.googleusercontent.com/VONu2F1QvtTPBBgHvDZQbZBuierz1MIHGoQ9_QRsGpiu4ulrLQqeemJTboUI5Pnlcg=h900</image>
+    </screenshot>
+  </screenshots>
+  <project_license>LicenseRef-proprietary</project_license>
+  <developer_name>Mojang AB</developer_name>
```
Comment (nedrichards) [inline com.mojang.Minecraft.appdata.xml]: We're encouraging people to add [OARS](https://odrs.gnome.org/oars) metadata, it's really easy!

**Comment by nedrichards** [comment]
The network issue is SSL related I think.

File: com.mojang.Minecraft.appdata.xml (hunk 4 above)
Comment (Kekun) [reply com.mojang.Minecraft.appdata.xml]: Thanks, I added them, I also put a proper description.

**Comment by nedrichards** [comment]
bot, build com.mojang.Minecraft

**Comment by Kekun** [comment]
32 bit ARM failed, so I added this to the repo: https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk9/blob/master/flathub.json

**Comment by nedrichards** [comment]
bot, build com.mojang.Minecraft

**Comment by Kekun** [comment]
Great! :) Know all we need is to understand why network doesn't work (I don't understand what you meant by mentioning SSL).

**Comment by nedrichards** [comment]
@Kekun that error message is talking about trustAnchors and invalid algorithms, which is all 'SSL talk' and a bit of idle googling around that suggests something about java's CA certificates (I think java has a different trust store to the main OS).

**Comment by Kekun** [comment]
I tried adding this to the finish-args, it didn't work:
```
"--env=JAVA_HOME=/app/jre",
"--env=JAVA_BINDIR=/app/jre/bin",
"--env=JAVA_ROOT=/app/jre"
```

**Comment by AsciiWolf** [comment]
It looks like that the OpenJDK trust store is missing or empty.
I have tried adding `-Djavax.net.ssl.trustStore=/app/jre/jvm/openjdk-9/lib/security/cacerts` to the Minecraft launch command, but unfortunately, no luck.
I believe Debian has _ca-certificates-java_ package for this, so maybe we could use it.

**Comment by nedrichards** [comment]
@AsciiWolf That sounds like the sort of thing every app would need. Should that be part of the JDK package?

**Comment by AsciiWolf** [comment]
@nedrichards Yep, I think so. Unfortunately, I don't have that much experience with Java and was not able to fix this.

**Comment by Tids** [comment]
@Kekun 
> required files in ~/.minecraft and then is fine. Removing ~/.minecraft

You can use 

```_JAVA_OPTIONS="-Duser.home=${XDG_DATA_HOME}/java"```

Minecraft respects this and will download the files to given directory. But still will still create a "`.minecraft`" folder in it. So minecraft directory in this case would be: `${XDG_DATA_HOME}/java/.minecraft`

**Comment by nedrichards** [comment]
bot, build com.mojang.Minecraft

**Comment by alexlarsson** [comment]
This fixes the certificate store, but it still fails to run, first because:
```
Unrecognized VM option 'CMSIncrementalMode'
```
But if i remove that:
```
Exception in thread "main" java.lang.UnsatisfiedLinkError: no lwjgl64 in java.library.path
```

There is also a warning:
```
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
```
Maybe MineCraft was designed for a different jdk version?

**Comment by alexlarsson** [comment]
Maybe we need to build the lwjgl library as a dependency?

**Comment by Alexander-Wilms** [comment]
Does changing to OpenJDK 10 improve things?

**Comment by nedrichards** [comment]
Wait, no - I got this working, will try and polish.

File: com.mojang.Minecraft.json [hunk 5]
```diff
@@ -66,7 +64,7 @@
         {
           "type": "script",
           "commands": [
-            "/app/jre/bin/java -Djavax.net.ssl.trustStorePassword=changeit -jar /app/extra/Minecraft.jar"
+            "/app/jre/bin/java -jar /app/extra/Minecraft.jar /app/jre/bin/java"
```
[the remark this answers is not in the document]
Comment (AsciiWolf) [reply com.mojang.Minecraft.json]: <pre>
diff --git a/com.mojang.Minecraft.json b/com.mojang.Minecraft.json
index cc7473b..3c75f13 100644
--- a/com.mojang.Minecraft.json
+++ b/com.mojang.Minecraft.json
@@ -64,7 +64,7 @@
         {
           "type": "script",
           "commands": [
-            "/app/jre/bin/java -jar /app/extra/Minecraft.jar /app/jre/bin/java"
+            "/app/jre/bin/java -jar /app/extra/Minecraft.jar"
           ],
           "dest-filename": "minecraft"
         },
</pre>
;-)

**Comment by AsciiWolf** [comment]
Minecraft 1.13 was [released](https://minecraft.net/en-us/article/update-aquatic-out-java) and it looks like it uses sandboxed Chrome for the launcher so the question is if it will be possible to get this working in Flatpak.

**Comment by nedrichards** [comment]
Oh! Interesting. Worth a look for sure.

**Comment by nedrichards** [comment]
@AsciiWolf so the launcher is just an electron app now, which is fine and very easy to get running. I get an error that my system can't find java though, which it thinks should be in /usr/bin/java - that's likely to be more of a problem.

**Comment by AsciiWolf** [comment]
@nedrichards Got it working! I have found out that the new launcher expects Java to be in `/app`, not `/app/jre`, so I have just re-added the Java-related stuff, then added a JAVA_HOME env variable pointing to the actual Java location. Minecraft now runs, singleplayer game seems to work without any issues, but multiplayer does not - connection to a server fails with "Failed to log in: The authentication servers are currently down for maintenance." and Minecraft Realms show "Exception: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException". Maybe another issue with `/app/jre/lib/security/cacerts`?

Here's the patch:
<pre>
diff --git a/com.mojang.Minecraft.json b/com.mojang.Minecraft.json
index 1f1979f..1c0c1f6 100644
--- a/com.mojang.Minecraft.json
+++ b/com.mojang.Minecraft.json
@@ -4,10 +4,14 @@
   "runtime-version": "1.6",
   "sdk": "org.freedesktop.Sdk",
   "command": "minecraft",
+  "sdk-extensions": [
+    "org.freedesktop.Sdk.Extension.openjdk10"
+  ],
   "base": "io.atom.electron.BaseApp",
   "base-version": "stable",
   "finish-args": [
     "--persist=.minecraft",
+    "--env=JAVA_HOME=/app/jre",
     /* X11 access */
     "--socket=x11",
     "--share=ipc",
@@ -18,6 +22,9 @@
     /* Network access */
     "--share=network"
   ],
+  "build-options" : {
+      "append-path": "/usr/lib/sdk/openjdk10/bin"
+  },
   "modules": [
     {
       /* Needed by Minecraft 1.8.2 and up. */
@@ -30,6 +37,13 @@
         }
       ]
     },
+    {
+      "name": "openjdk",
+      "buildsystem": "simple",
+      "build-commands": [
+        "/usr/lib/sdk/openjdk10/install.sh"
+      ]
+    },
     {
       "name": "minecraft",
       "buildsystem": "simple",
</pre>

**Comment by nedrichards** [comment]
@AsciiWolf thanks! I've added that.

**Comment by AsciiWolf** [comment]
I have [made](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk10/issues/4#issuecomment-421941449) some progress, however the multiplayer still doesn't work. I also noticed that there is one more issue that needs to be resolved: The Minecraft launcher process does not exit properly and is still running in background after the game window is closed.

**Comment by AsciiWolf** [comment]
I have updated the json manifest to use OpenJDK 11:
<pre>
diff --git a/com.mojang.Minecraft.json b/com.mojang.Minecraft.json
index 45f9f68..1dd10d0 100644
--- a/com.mojang.Minecraft.json
+++ b/com.mojang.Minecraft.json
@@ -1,13 +1,13 @@
 {
   "app-id": "com.mojang.Minecraft",
   "runtime": "org.freedesktop.Platform",
-  "runtime-version": "1.6",
+  "runtime-version": "18.08",
   "sdk": "org.freedesktop.Sdk",
   "command": "minecraft",
   "base": "io.atom.electron.BaseApp",
   "base-version": "stable",
   "sdk-extensions": [
-    "org.freedesktop.Sdk.Extension.openjdk10"
+    "org.freedesktop.Sdk.Extension.openjdk11"
   ],
   "finish-args": [
     "--persist=.minecraft",
@@ -23,7 +23,7 @@
     "--env=JAVA_HOME=/app/jre"
   ],
   "build-options" : {
-    "append-path": "/usr/lib/sdk/openjdk10/bin"
+    "append-path": "/usr/lib/sdk/openjdk11/bin"
   },
   "modules": [
     {
@@ -41,7 +41,7 @@
       "name": "openjdk",
         "buildsystem": "simple",
         "build-commands": [
-          "/usr/lib/sdk/openjdk10/install.sh"
+          "/usr/lib/sdk/openjdk11/install.sh"
         ]
     },
     {
@@ -61,8 +61,8 @@
           "type": "extra-data",
           "filename": "Minecraft.tar.gz",
           "url": "https://launcher.mojang.com/download/Minecraft.tar.gz",
-          "sha256": "a23e0df912f3f3772f8ad376ef9237a02264327e3048435897fdf7d4e71c549b",
-          "size": 57828753
+          "sha256": "b0026863e8a7d41d328d831aa223fdb198a96a6bcf980969a7dc065ed29a4f06",
+          "size": 57840610
         },
         {
           "type": "script",
</pre>

Seems to work fine, except the multiplayer, but it is a different error now: "Could not generate ECDH keypair" - that is a good sign because it is [this bug](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk11/issues/2) that will hopefully be fixed. Oh, and the game still hangs on exit, maybe related to this error: `AL lib: (EE) alc_cleanup: 1 device not closed`

**Comment by mbooth101** [comment]
Hi, can you try rebuilding against the latest build of OpenJDK 11?

I this SSL problems should be gone now.

**Comment by AsciiWolf** [comment]
I can confirm that multiplayer now works without any issues! Unfortunately, there is still a problem with the game process that hangs on exit.

**Comment by mbooth101** [comment]
Great, thanks for testing :-)

**Comment by nedrichards** [comment]
bot, build com.mojang.Minecraft

**Comment by AsciiWolf** [comment]
I have tried to debug the issue with the game hanging on quit. It looks like that it's the launcher hanging, not the game itself. And it doesn't happen every time. Here are some strace logs, unfortunately not a much relevant information in them:

[minecraft_launcher.bad.log](https://github.com/flathub/flathub/files/2514289/minecraft_launcher.bad.log) (when the game hangs)
[minecraft_launcher.good.log](https://github.com/flathub/flathub/files/2514290/minecraft_launcher.good.log) (when the game exits correctly)

The only interesting thing in them is this:

`wait4(-1, 0x7ffebd5bb900, 0, NULL)      = ? ERESTARTSYS (To be restarted if SA_RESTART is set)` // when the game hangs

`wait4(-1, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], 0, NULL) = 2892` // when the game exits correctly

**Comment by AsciiWolf** [comment]
Maybe this is a bug in Electron itself.

**Comment by Kekun** [comment]
@AsciiWolf Not from my side, I didn't touch this since ages because the encountered problems go *way* beyond my understanding of Java and networking. :x I'm happy to leave that to whoever wants to take it over.

**Comment by AsciiWolf** [comment]
@Kekun The network issues are already fixed an everything in the game is fully working. :-) The only remaining issue is that the game process hangs on exit for some reason.

**Comment by TLATER** [comment]
@AsciiWolf Could this possibly be merged with that bug? There would more eyes on the issue that way, and it seems relatively minor.

**Comment by AsciiWolf** [comment]
@TLATER I don't think it is a good idea to merge this with the bug. It could seem like a minor issue, but the fact that almost every Minecraft instance keeps running on background forever seems like a big problem to me because it can cause slowdowns, crashes and other issues.

**Comment by yodatak** [comment]
I try to find something on their bug tracker maybe i can be interesting to talk to them by twitter or by their bug tracker ? 
https://bugs.mojang.com/browse/MC-114835?jql=text%20~%20%22exit%22

**Comment by AsciiWolf** [comment]
New update:
```
diff --git a/com.mojang.Minecraft.json b/com.mojang.Minecraft.json
index 1dd10d0..76abba1 100644
--- a/com.mojang.Minecraft.json
+++ b/com.mojang.Minecraft.json
@@ -61,22 +61,21 @@
           "type": "extra-data",
           "filename": "Minecraft.tar.gz",
           "url": "https://launcher.mojang.com/download/Minecraft.tar.gz",
-          "sha256": "b0026863e8a7d41d328d831aa223fdb198a96a6bcf980969a7dc065ed29a4f06",
-          "size": 57840610
+          "sha256": "06e7649f702b782fd606361a18ce133e108cc33a11a800abea0ba5f4caba2fe6",
+          "size": 58188591
         },
         {
           "type": "script",
           "dest-filename": "apply_extra",
           "commands": [
               "tar -xf Minecraft.tar.gz",
-              "rm -f Minecraft.tar.gz",
-              "mv minecraft-launcher-* minecraft-launcher"
+              "rm -f Minecraft.tar.gz"
           ]
         },
         {
           "type": "script",
           "commands": [
-            "/app/extra/minecraft-launcher/minecraft-launcher.sh"
+            "/app/extra/minecraft-launcher/minecraft-launcher"
           ],
           "dest-filename": "minecraft"
         },
```

And I have fantastic news! Minecraft no longer hangs on exit! I have been playing it for last hour and everything seems to work perfectly. Both singleplayer and multiplayer. And no hang when I exit the game. :-) That means that the Minecraft Flatpak is **no longer blocked** from being published on Flathub.

**Comment by LeandroStanger** [comment]
bot, build com.mojang.Minecraft

**Comment by nedrichards** [comment]
@AsciiWolf great news! let's get that in there and tested ASAP.

**Comment by AsciiWolf** [comment]
@nedrichards Thanks! Don't forget to apply the patch that I provided in my previous comment. :-)

**Comment by nedrichards** [comment]
Indeed. I'll try and take a look at it tomorrow.

**Comment by nedrichards** [comment]
bot, build com.mojang.Minecraft

**Comment by nedrichards** [comment]
bot, build com.mojang.Minecraft

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/com.mojang.Minecraft

**Pull request closed without merging**